source/extensions/io_socket/io_uring/io_handle_impl.cc

```diff
@@ -0,0 +1,365 @@
+                                                         socklen_t* optlen) {
+  return Api::OsSysCallsSingleton::get().getsockopt(fd_, level, optname, optval, optlen);
+}
+
+Api::SysCallIntResult IoUringSocketHandleImpl::ioctl(unsigned long, void*, unsigned long, void*,
+                                                     unsigned long, unsigned long*) {
+  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+}
+
+Api::SysCallIntResult IoUringSocketHandleImpl::setBlocking(bool /*blocking*/) {
+  NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
+}
+
+absl::optional<int> IoUringSocketHandleImpl::domain() { return domain_; }
+
+Network::Address::InstanceConstSharedPtr IoUringSocketHandleImpl::localAddress() {
+  // TODO(rojkov): This is a copy-paste from Network::IoSocketHandleImpl.
+  // Unification is needed.
+  sockaddr_storage ss;
+  socklen_t ss_len = sizeof(ss);
+  auto& os_sys_calls = Api::OsSysCallsSingleton::get();
+  Api::SysCallIntResult result =
+      os_sys_calls.getsockname(fd_, reinterpret_cast<sockaddr*>(&ss), &ss_len);
+  if (result.return_value_ != 0) {
+    throw EnvoyException(fmt::format("getsockname failed for '{}': ({}) {}", fd_, result.errno_,
+                                     errorDetails(result.errno_)));
+  }
+  return Network::Address::addressFromSockAddrOrThrow(ss, ss_len, socket_v6only_);
+}
+
+Network::Address::InstanceConstSharedPtr IoUringSocketHandleImpl::peerAddress() {
+  // TODO(rojkov): This is a copy-paste from Network::IoSocketHandleImpl.
+  // Unification is needed.
+  sockaddr_storage ss;
+  socklen_t ss_len = sizeof ss;
+  auto& os_sys_calls = Api::OsSysCallsSingleton::get();
+  Api::SysCallIntResult result =
+      os_sys_calls.getpeername(fd_, reinterpret_cast<sockaddr*>(&ss), &ss_len);
+  if (result.return_value_ != 0) {
+    throw EnvoyException(
+        fmt::format("getpeername failed for '{}': {}", errorDetails(result.errno_)));
+  }
+
+  if (ss_len == udsAddressLength() && ss.ss_family == AF_UNIX) {
+    // For Unix domain sockets, can't find out the peer name, but it should match our own
+    // name for the socket (i.e. the path should match, barring any namespace or other
+    // mechanisms to hide things, of which there are many).
+    ss_len = sizeof ss;
+    result = os_sys_calls.getsockname(fd_, reinterpret_cast<sockaddr*>(&ss), &ss_len);
+    if (result.return_value_ != 0) {
+      throw EnvoyException(
+          fmt::format("getsockname failed for '{}': {}", fd_, errorDetails(result.errno_)));
+    }
+  }
+  return Network::Address::addressFromSockAddrOrThrow(ss, ss_len, socket_v6only_);
+}
+
+void IoUringSocketHandleImpl::initializeFileEvent(Event::Dispatcher& dispatcher,
+                                                  Event::FileReadyCb cb,
+                                                  Event::FileTriggerType trigger, uint32_t events) {
+  // Check if this is a server socket accepting new connections.
+  if (isLeader()) {
```

Comment:
under which scenario is `isLeader()` `true` in `initializeFileEvent`?

Is there a case `file_event_adapter_` is not a nullptr and the caller wants to `initializeFileEvent`?

Reply:
> under which scenario is `isLeader()` `true` in `initializeFileEvent`?
> Is there a case `file_event_adapter_` is not a `nullptr` and the caller wants to `initializeFileEvent`?

Only in case of a listener socket. In this case `file_event_adapter_` is set in `listen()` which is called before `initializeFileEvent()`. For client sockets `listen()` is never called.